File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        mp = {'(': ')', '{': '}', '[': ']'}
    
        stack = []
        
        for char in s:
            # If it's an opening bracket, push it to the stack
            if char in mp:
                stack.append(char)
            # If it's a closing bracket, check for matching opening bracket
            elif stack and mp[stack[-1]] == char:
                stack.pop()
            # If no match, return false
            else:
                return False
        
        # Return true if stack is empty
        return not stack

File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;
        unordered_map<char, char> map = {
            {')','('},
            {'}','{'},
            {']','['}
        };

        for(char i : s){
            if(map.find(i) == map.end()){
                stk.push(i);
            }
            else if(!stk.empty() && stk.top() == map[i]){
                stk.pop();
            }
            else{
                return false;
            }
        }
        return stk.empty();
    }
};